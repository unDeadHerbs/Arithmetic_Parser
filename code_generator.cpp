#include "code_generator.hpp"
#include <stdlib.h>
#include <sys/mman.h>
#include <numeric>
using std::accumulate;
using namespace std::string_literals;

namespace {
#include <iostream>
#include <string>
void print_int(int x) { std::cout << x; }
void print_str(char* s) { std::cout << s; }
void print_bool(int b) { std::cout << (b ? "true" : "false"); }
int read_int() {
	int x;
	std::cin >> x;
	return x;
}
}  // namespace

void program::operator()() {
	auto prog =
	    (char*)mmap(NULL /*we don't care where we point to.*/,
	                buffer.size() /*size of buffer*/,
	                PROT_EXEC | PROT_READ | PROT_WRITE /*permissions*/,
	                MAP_PRIVATE | MAP_ANONYMOUS /**/,
	                -1 /*ignored as not a file*/, 0 /*ignored as not a file*/);
	if (!prog) throw "Failed to alocate space"s;
	int end(0);
	for (auto c : buffer) prog[end++] = c;
	((int (*)())prog)();
	munmap(prog, buffer.size());
}
#define LITTLEENDIAN4(x)                                                       \
	program({(char)(((x) >> 0x08 * 0) & 0xFF), (char)(((x) >> 0x08 * 1) & 0xFF), \
	         (char)(((x) >> 0x08 * 2) & 0xFF),                                   \
	         (char)(((x) >> 0x08 * 3) & 0xFF)})
#define LITTLEENDIAN8(x)                                                       \
	program({(char)(((x) >> 0x08 * 0) & 0xFF), (char)(((x) >> 0x08 * 1) & 0xFF), \
	         (char)(((x) >> 0x08 * 2) & 0xFF), (char)(((x) >> 0x08 * 3) & 0xFF), \
	         (char)(((x) >> 0x08 * 4) & 0xFF), (char)(((x) >> 0x08 * 5) & 0xFF), \
	         (char)(((x) >> 0x08 * 6) & 0xFF),                                   \
	         (char)(((x) >> 0x08 * 7) & 0xFF)})
#define _MOVI_EAX program({((char)(0xB8 + 0x0))})

#define ADD_EAX_EDX program({((char)(0x01 + 0x00)), ((char)(0xd0))})
#define SUB_EAX_EDX program({((char)(0x29 + 0x00)), ((char)(0xd0))})
#define MUL_EAX_EDX program({((char)(0x0F)), ((char)(0xAF)), ((char)(0xC2))})
#define DIV_EDAX_ECX program({((char)(0xF7)), ((char)(0xF9))})

#define PUSH_EAX program({((char)(0x50 + 0x00))})
#define PUSH_ECX program({((char)(0x50 + 0x01))})
#define PUSH_EDX program({((char)(0x50 + 0x02))})
#define POP_EAX program({((char)(0x58 + 0x00))})
#define POP_ECX program({((char)(0x58 + 0x01))})
#define POP_EDX program({((char)(0x58 + 0x02))})
#define POP_RDI program({((char)(0x58 + 0x07))})
#define MOV_RDI_RAX program({((char)0x89), ((char)0xc7)})
#define MOD_EDAX_ECX DIV_EDAX_ECX + PUSH_EDX + POP_EAX
#define LOADI(x) _MOVI_EAX + LITTLEENDIAN4((int)(x))
#define ZERO_EAX LOADI(0)
#define LOADI_S(x) LOADI(strtol(x.c_str(), NULL, 10))
#define _MOVI_RSI program({((char)(0x48)), ((char)(0xbe))})
#define _MOVI_RDI program({((char)(0x48)), ((char)(0xbf))})
#define _MOVI_RAX program({((char)(0x48)), ((char)(0xc8))})
#define LOADLL_RDI(x) _MOVI_RDI + LITTLEENDIAN8((x))
#define LOADLL_RAX(x) _MOVI_RAX + LITTLEENDIAN8((x))
#define LOADLL_STR(x) _MOVI_RDI + LITTLEENDIAN8((x))
#define LOADLL_RSI(x) _MOVI_RSI + LITTLEENDIAN8((x))
#define CALL_RSI program({((char)(0xff)), ((char)(0xd6))})
#define PRINT_EAX_I \
	MOV_RDI_RAX + LOADLL_RSI((long long int)print_int) + CALL_RSI
#define PRINT_EAX_BOOL \
	MOV_RDI_RAX + LOADLL_RSI((long long int)print_bool) + CALL_RSI
#define PRINT_STR LOADLL_RSI((long long int)print_str) + CALL_RSI
#define READ_EAX LOADLL_RSI((long long int)read_int) + CALL_RSI
#define MOV_EAX_STACK(d) \
	program({((char)0x8b), ((char)0x45), ((char)((-4 * (d + 1)) & 0xff))})
#define MOV_STACK_EAX(d) \
	program({((char)0x89), ((char)0x45), ((char)((-4 * (d + 1)) & 0xff))})
// mov eax,[ebp-4*d]
#define _BOOL_AL_TO_EAX \
	program({((char)0xc0), ((char)0x66), ((char)0x98), ((char)0x98)})
#define CMP_EAX_ECX program({((char)0x39), ((char)0xc8)})
#define BOOL_IS_LESS program({((char)0x0f), ((char)0x9c)}) + _BOOL_AL_TO_EAX
#define BOOL_EAX_LESS_ECX CMP_EAX_ECX + BOOL_IS_LESS
#define BOOL_IS_LESS_EQ program({((char)0x0f), ((char)0x9e)}) + _BOOL_AL_TO_EAX
#define BOOL_IS_EQ program({((char)0x0f), ((char)0x94)}) + _BOOL_AL_TO_EAX
#define BOOL_IS_GREATER_EQ \
	program({((char)0x0f), ((char)0x9d)}) + _BOOL_AL_TO_EAX
#define BOOL_IS_GREATER program({((char)0x0f), ((char)0x9f)}) + _BOOL_AL_TO_EAX
#define BOOL_IS_NE program({((char)0x0f), ((char)0x95)}) + _BOOL_AL_TO_EAX
#define BOOL_EAX_OR_ECX program({((char)0x09), ((char)0xc8)})
#define BOOL_EAX_AND_ECX program({((char)0x21), ((char)0xc8)})
//#define BOOL_NOT program({((char)0xf7), ((char)0xd0)})
#define BOOL_NOT PUSH_EAX + LOADI(1) + POP_EDX + SUB_EAX_EDX
// TODO: make that not command much more effecent

program generateI(Code_Tree ct, std::shared_ptr<Env> e) {
	// TODO: Those throws should get the TOKEN's name.
	if (ct.name == "numeric_literal") {
		if (ct.t->id == INT) return LOADI_S(ct.t->text);
		throw "Unsupported Literal Type"s;
	}
	if (ct.name == "bool_literal") {
		if (ct.t->text == "true") return LOADI(1);
		if (ct.t->text == "false") return LOADI(0);
		throw "Unsupported bool literal '"s + ct.t->text + "'";
	}
	if (ct.name == "add") {
		auto p1 = generateI(ct.sub_tokens[0], e);
		auto p2 = generateI(ct.sub_tokens[1], e);
		if (ct.t->id == '+') return p2 + PUSH_EAX + p1 + POP_EDX + ADD_EAX_EDX;
		if (ct.t->id == '-') return p2 + PUSH_EAX + p1 + POP_EDX + SUB_EAX_EDX;
		throw "Unknown add OP "s + (char)ct.t->id;
	}
	if (ct.name == "mul") {
		auto p1 = generateI(ct.sub_tokens[0], e);
		auto p2 = generateI(ct.sub_tokens[1], e);
		if (ct.t->id == '*') return p2 + PUSH_EAX + p1 + POP_EDX + MUL_EAX_EDX;
		if (ct.t->id == '/')
			// TODO: properly handel sign extention
			return ZERO_EAX + PUSH_EAX + p2 + PUSH_EAX + p1 + POP_ECX + POP_EDX +
			       DIV_EDAX_ECX;
		if (ct.t->text == "mod")
			return ZERO_EAX + PUSH_EAX + p2 + PUSH_EAX + p1 + POP_ECX + POP_EDX +
			       MOD_EDAX_ECX;
		throw "Unknown mul OP "s + (char)ct.t->id;
	}
	if (ct.name == "exp") {
		if (ct.t->id == '^') return generateI(ct.sub_tokens[0], e);
		// TODO: impliment this
		throw "Unknown exp OP "s + (char)ct.t->id;
	}
	if (ct.name == "unary") {
		auto p = generateI(ct.sub_tokens[0], e);
		if (ct.t->id == '+') return p;
		if (ct.t->id == '-') return p + PUSH_EAX + ZERO_EAX + POP_EDX + SUB_EAX_EDX;
		throw "Unknown Unary Op "s + (char)ct.t->id;
	}
	if (ct.name == "block") {
		return accumulate(
		    ct.sub_tokens.begin(), ct.sub_tokens.end(), program(),
		    [&e](program p, Code_Tree c) { return p + generateI(c, e); });
	}
	if (ct.name == "print") {
		auto ret = program();
		for (auto c : ct.sub_tokens)
			if (c.name == "print_i")
				ret += generateI(c.sub_tokens[0], e) + PRINT_EAX_I;
			else if (c.name == "print_s") {
				auto p = e->push(c.sub_tokens[0].t->text);
				ret += LOADLL_STR((long long int)p) + PRINT_STR;
			} else if (c.name == "print_b") {
				ret += generateI(c.sub_tokens[0], e) + PRINT_EAX_BOOL;
			} else
				throw "Unknown print mode "s + c.name;
		return ret;
	}
	if (ct.name == "Declaration") {
		e->cleanup.push_back(POP_EAX);
		e->syms.push_back(ct.sub_tokens[0].t->text);
		return PUSH_EAX;
	}
	if (ct.name == "Ident") {
		auto var = ct.t->text;
		return MOV_EAX_STACK(e->lookup(var));
	}
	if (ct.name == "Read") {
		auto var = ct.sub_tokens[0].t->text;
		return READ_EAX + MOV_STACK_EAX(e->lookup(var));
	}
	if (ct.name == "Assignment") {
		auto var = ct.sub_tokens[0].t->text;
		return generateI(ct.sub_tokens[1], e) + MOV_STACK_EAX(e->lookup(var));
	}
	if (ct.name == "bool_rel") {
		auto ret = generateI(ct.sub_tokens[2], e) + PUSH_EAX +
		           generateI(ct.sub_tokens[1], e) + POP_ECX + CMP_EAX_ECX;
		if (ct.sub_tokens[0].t->id == '<')
			return ret + BOOL_IS_LESS;
		else if (ct.sub_tokens[0].t->id == LESS_EQ)
			return ret + BOOL_IS_LESS_EQ;
		else if (ct.sub_tokens[0].t->id == '=')
			return ret + BOOL_IS_EQ;
		else if (ct.sub_tokens[0].t->id == GREATER_EQ)
			return ret + BOOL_IS_GREATER_EQ;
		else if (ct.sub_tokens[0].t->id == '>')
			return ret + BOOL_IS_GREATER;
		else if (ct.sub_tokens[0].t->id == NOT_EQUAL)
			return ret + BOOL_IS_NE;
		else
			throw "Expected a boolean realational, got '"s +
			    ct.sub_tokens[0].t->text + "'";
	}
	if (ct.name == "bool_or") {
		return generateI(ct.sub_tokens[1], e) + PUSH_EAX +
		       generateI(ct.sub_tokens[0], e) + POP_ECX + BOOL_EAX_OR_ECX;
	}
	if (ct.name == "bool_and") {
		return generateI(ct.sub_tokens[1], e) + PUSH_EAX +
		       generateI(ct.sub_tokens[0], e) + POP_ECX + BOOL_EAX_AND_ECX;
	}
	if (ct.name == "bool_not") {
		return generateI(ct.sub_tokens[0], e) + BOOL_NOT;
	}
	throw "Bad Parse Tree - unknown '"s + ct.name + "'";
}
#define RETURN ((char)(0xC3))
program generate(Code_Tree ct) {
	program r;
	r.init();
	r += generateI(ct, r.e);
	r.cleanup();
	return r + RETURN;
}
