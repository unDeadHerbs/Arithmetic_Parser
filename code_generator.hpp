#ifndef __CODE_GENERATOR_HPP__
#define __CODE_GENERATOR_HPP__
#include "code_tree.hpp"
#include <utility>
#include <string.h>
#include <memory>

typedef unsigned char binary;

struct sym_props{};
struct program;
struct Env{
  char string_table[5000];
  char* end;
  std::vector<program> cleanup;
  std::vector<std::pair<std::string,sym_props>> syms;
public:
  Env():string_table{0},end(string_table){}
  char* push(std::string s){
    strcpy(end,s.c_str());
    auto ret=end;
    while(*end++);
    return ret;
  }
};

struct program{
  std::vector<char> buffer;
  std::shared_ptr<Env> e;
public:
  program(){}
  program(std::vector<char>b):buffer(b){}
  void operator()(); //TODO: add optional parameters
  void init(){e=std::make_shared<Env>();}
  void cleanup(){for(auto &c:e->cleanup)operator+=(c);}
  int size(){return buffer.size();}
  program& operator+=(program p){
    for(auto c:p.buffer)
      buffer.push_back(c);
    return *this;
  }
  program& operator+=(char c){
    buffer.push_back(c);
    return *this;
  }
  program operator+(program p){
    program t(*this);
    return t+=p;
  }
  program operator+(char c){
    program t(*this);
    return t+=c;
  }
  //operator += vec<opcode>
  //operator [] for indexing
  //Add an insert? Nope that's hard.
};
program generate(Code_Tree);

#endif
