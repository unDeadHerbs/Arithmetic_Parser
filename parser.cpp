#include "parser.hpp"
using std::vector;

Code_Tree parse_expression(vector<Token>& tokens);

#define _FUN std::string(__func__)
#define EOF_ERROR(x) Code_Tree(Token(ERROR, -1, -1, x))
#define ERROR(x, t) \
	Code_Tree("ERROR", Token(ERROR, t.line, t.col, x), {Code_Tree(t)})
#define EAT(t, n) t.erase(t.begin(), t.begin() + n)
#define EAT_OP(t, tokens, op)                                            \
	if (!tokens.size()) throw EOF_ERROR(_FUN + " Expected '" + #op + "'"); \
	auto t = tokens[0];                                                    \
	if (t.id != op) throw ERROR(_FUN + " Expected '" + #op + "'", t);      \
	EAT(tokens, 1);

#define EAT_IDENT(t, tokens, str)                                        \
	if (!tokens.size()) throw EOF_ERROR(_FUN + " Expected '" + str + "'"); \
	auto t = tokens[0];                                                    \
	if (t.id != IDENT) throw ERROR(_FUN + " Expected '" + str + "'", t);   \
	if (t.text != str) throw ERROR(_FUN + " Expected '" + str + "'", t);   \
	EAT(tokens, 1);

#define EAT_STRING(t, tokens)                                       \
	if (!tokens.size()) throw EOF_ERROR(_FUN + " Expected 'string'"); \
	auto t = tokens[0];                                               \
	if (t.id != STRING) throw ERROR(_FUN + " Expected 'string'", t);  \
	EAT(tokens, 1);

#define is_keyword(t) \
	(((t) == "print") || ((t) == "read") || ((t) == "true") || ((t) == "false"))

Code_Tree parse_ident(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("Expected variable name");
	auto t = tokens[0];
	if (t.id != IDENT) throw ERROR("Expected identifier name", t);
	if (is_keyword(t.text)) throw ERROR("Keywords are not identifiers", t);
	EAT(tokens, 1);
	return Code_Tree("Ident", t, {});
}

Code_Tree parse_number(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto t = tokens[0];
	if (t.id == INT || t.id == REAL) {
		EAT(tokens, 1);
		return Code_Tree("numeric_literal", t);
	}
	throw ERROR("Not a Litteral", t);
}

Code_Tree parse_value(vector<Token>& tokens) {
	auto toks = tokens;
	try {
		EAT_OP(t, toks, '(');
		auto ret = parse_expression(toks);
		EAT_OP(t1, toks, ')');
		tokens = toks;
		return ret;
	} catch (Code_Tree err1) {
		try {
			auto ret = parse_number(toks);
			tokens = toks;
			return ret;
		} catch (Code_Tree err2) {
			try {
				auto ret = parse_ident(toks);
				tokens = toks;
				return ret;
			} catch (Code_Tree err3) {
				throw Code_Tree(
				    "Expected a parenthetical, number, or ident.",
				    Token(ERROR, -1, -1, "Expected a parenthetical, number, or ident."),
				    {err1, err2, err3});
			}
		}
	}
}

Code_Tree parse_unary_signs(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto t = tokens[0];
	if (t.id != '+' && t.id != '-') return parse_value(tokens);
	EAT(tokens, 1);
	return Code_Tree("unary", t, {parse_value(tokens)});
}

Code_Tree parse_exp(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto ret = parse_unary_signs(tokens);
	if (!tokens.size()) return ret;
	auto t = tokens[0];
	if (t.id != '^') return ret;
	EAT(tokens, 1);
	return Code_Tree("exp", t, {ret, parse_exp(tokens)});
}

Code_Tree parse_mul_div_prime(vector<Token>& tokens, Code_Tree accumulator) {
	if (!tokens.size()) return accumulator;
	auto t = tokens[0];
	if (!(t.id == '*' || t.id == '/' || t.text == "mod")) return accumulator;
	EAT(tokens, 1);
	return parse_mul_div_prime(
	    tokens, Code_Tree("mul", t, {accumulator, parse_exp(tokens)}));
}

Code_Tree parse_mul_div(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto ret = parse_exp(tokens);
	return parse_mul_div_prime(tokens, ret);
}

Code_Tree parse_sum_dif_prime(vector<Token>& tokens, Code_Tree accumulator) {
	if (!tokens.size()) return accumulator;
	auto t = tokens[0];
	if (!(t.id == '+' || t.id == '-')) return accumulator;
	EAT(tokens, 1);
	return parse_sum_dif_prime(
	    tokens, Code_Tree("add", t, {accumulator, parse_mul_div(tokens)}));
}

Code_Tree parse_sum_dif(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto ret = parse_mul_div(tokens);
	return parse_sum_dif_prime(tokens, ret);
}

Code_Tree parse_expression(vector<Token>& tokens) {
	return parse_sum_dif(tokens);
}

Code_Tree parse_string(vector<Token>& tokens) {
	EAT_STRING(t, tokens);
	return Code_Tree("String", t);
}

Code_Tree parse_bool_relation(vector<Token>& tokens) {
	auto fir = parse_expression(tokens);
	auto nzret = Code_Tree("bool_imp", {fir});
	if (!tokens.size()) return nzret;
	auto t = tokens[0];
	if (!((t.id == '<') || (t.id == LESS_EQ) || (t.id == '>') ||
	      (t.id == GREATER_EQ) || (t.id == '=') || (t.id == NOT_EQUAL)))
		return nzret;
	EAT(tokens, 1);
	try {
		auto sec = parse_expression(tokens);  // then eat arith
		return Code_Tree("bool_rel", {fir, Code_Tree(t), sec});
	} catch (Code_Tree err) {
		throw Code_Tree(_FUN + " Expected an arithmetic expresion while parsing",
		                {fir, Code_Tree(t), err});
	}
}

Code_Tree parse_bool_literal(vector<Token>& tokens) {
	try {
		auto tmp = tokens;
		EAT_IDENT(t, tmp, "true");
		tokens = tmp;
		return t;
	} catch (Code_Tree err1) {
		try {
			auto tmp = tokens;
			EAT_IDENT(t, tmp, "false");
			tokens = tmp;
			return t;
		} catch (Code_Tree err2) {
			throw Code_Tree("Expected true or false", {err1, err2});
		}
	}
}

Code_Tree parse_bool(vector<Token>& tokens) {
	try {
		auto tmp = tokens;
		auto ret = parse_bool_literal(tmp);
		tokens = tmp;
		return ret;
	} catch (Code_Tree err1) {
		try {
			auto tmp = tokens;
			auto ret = parse_bool_relation(tmp);
			tokens = tmp;
			return ret;
		} catch (Code_Tree err2) {
			throw Code_Tree("Expected a boolean expresion", {err1, err2});
		}
	}
}

Code_Tree parse_printable(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto t = tokens[0];
	try {
		auto toks = tokens;
		auto ret = Code_Tree("print_i", {parse_expression(toks)});
		tokens = toks;
		return ret;
	} catch (Code_Tree err) {
		try {
			auto toks = tokens;
			auto ret = Code_Tree("print_s", {parse_string(toks)});
			tokens = toks;
			return ret;
		} catch (Code_Tree err2) {
			try {
				auto toks = tokens;
				auto ret = Code_Tree("print_b", {parse_bool(toks)});
				tokens = toks;
				return ret;
			} catch (Code_Tree err3) {
				throw Code_Tree(
				    "Expected expression, string, or bool.",
				    Token(ERROR, t.line, t.col, "Expected expression or string."),
				    {err, err2, err3});
			}
		}
	}
}

vector<Code_Tree> parse_printables_prime(vector<Token>& tokens,
                                         vector<Code_Tree> accumulator) {
	try {
		EAT_OP(t, tokens, ',');
		auto toks = tokens;
		accumulator.push_back(parse_printable(toks));
		tokens = toks;
		return parse_printables_prime(tokens, accumulator);
	} catch (Code_Tree err) {
		return accumulator;
	}
}

vector<Code_Tree> parse_printables(vector<Token>& tokens) {
	return parse_printables_prime(tokens, {parse_printable(tokens)});
}

Code_Tree parse_print_f(vector<Token>& tokens) {
	EAT_IDENT(t1, tokens, "print");
	EAT_OP(t2, tokens, '(');
	auto ret = parse_printables(tokens);
	EAT_OP(t3, tokens, ')');
	EAT_OP(t4, tokens, ';');
	return Code_Tree("print", t1, ret);
}

Code_Tree parse_read_f(vector<Token>& tokens) {
	EAT_IDENT(t1, tokens, "read");
	EAT_OP(t2, tokens, '(');
	auto ret = parse_ident(tokens);
	EAT_OP(t3, tokens, ')');
	EAT_OP(t4, tokens, ';');
	return Code_Tree("Read", t1, {ret});
}

Code_Tree parse_func_f(vector<Token>& tokens) {
	if (!tokens.size())
		throw EOF_ERROR("Expected a call to a function or function-like builtin");
	auto t = tokens[0];
	auto toks = tokens;
	try {
		auto ret = parse_print_f(toks);
		tokens = toks;
		return ret;
	} catch (Code_Tree err1) {
		try {
			auto ret = parse_read_f(toks);
			tokens = toks;
			return ret;
		} catch (Code_Tree err2) {
			throw Code_Tree(
			    "Expected a call to a function or function-like builtin",
			    Token(ERROR, t.line, t.col,
			          "Expected a call to a function or function-like builtin"),
			    {err1, err2});
		}
	}
}

Code_Tree parse_declare(vector<Token>& tokens) {
	EAT_IDENT(t1, tokens, "int4");
	auto ret = parse_ident(tokens);
	EAT_OP(t2, tokens, ';');
	return Code_Tree("Declaration", t1, {ret});
}

Code_Tree parse_assignment(vector<Token>& tokens) {
	auto var = parse_ident(tokens);
	EAT_OP(t1, tokens, ASSIGN);
	auto val = parse_expression(tokens);
	EAT_OP(t2, tokens, ';');
	return Code_Tree("Assignment", t1, {var, val});
}

Code_Tree parse_statement(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("Expected variable name");
	auto t = tokens[0];
	auto toks = tokens;
	try {
		auto ret = parse_func_f(toks);
		tokens = toks;
		return ret;
	} catch (Code_Tree err) {
		try {
			auto ret = parse_declare(toks);
			tokens = toks;
			return ret;
		} catch (Code_Tree err2) {
			try {
				auto ret = parse_assignment(toks);
				// TODO: move assignment to being a normal operator rather than
				// a builtin
				tokens = toks;
				return ret;
			} catch (Code_Tree err3) {
				throw Code_Tree(
				    "Expected a function, decleration, or assignment.",
				    Token(ERROR, t.line, t.col,
				          "Expected a function, decleration, or assignment."),
				    {err, err2, err3});
			}
		}
	}
}

vector<Code_Tree> parse_statements(vector<Token>& tokens,
                                   vector<Code_Tree> accumulator) {
	if (!tokens.size()) return accumulator;
	auto t = tokens[0];
	if (t.id == '}') return accumulator;
	accumulator.push_back(parse_statement(tokens));
	return parse_statements(tokens, accumulator);
}

Code_Tree parse_global_block(vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROF");
	return Code_Tree("block", parse_statements(tokens, {}));
}
Code_Tree parse_block(vector<Token>& tokens) {
	EAT_OP(t1, tokens, '{');
	auto ret = parse_global_block(tokens);
	EAT_OP(t2, tokens, '}');
	return Code_Tree("block", t1, {ret});
}

Code_Tree parse(vector<Token>& tokens) {
	try {
		return parse_global_block(tokens);
	} catch (Code_Tree err) {
		if (tokens.size()) {
			auto t = tokens[0];
			return Code_Tree("Parsing Error at " + std::to_string(t.line) + ':' +
			                     std::to_string(t.col),
			                 Token(ERROR, t.line, t.col,
			                       "Parsing Error at " + std::to_string(t.line) +
			                           ':' + std::to_string(t.col)),
			                 {err});
		}
		return Code_Tree("Parsing Error - EOF",
		                 Token(ERROR, -1, -1, "Parsing Error - EOF"), {err});
	}
}
