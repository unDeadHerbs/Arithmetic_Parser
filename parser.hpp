#ifndef __PARSER_HPP__
#define __PARSER_HPP__

#include <string>
#include <variant>
#include <vector>
#include "Tokenizer/tokenizer.hpp"
#include "code_tree.hpp"

Code_Tree parse(std::vector<Token>&);

#endif
