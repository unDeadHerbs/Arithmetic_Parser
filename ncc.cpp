#include <iostream>
using std::cout;
using std::endl;
#include "Tokenizer/tokenizer.hpp"
#include "code_generator.hpp"
#include "parser.hpp"
using std::ostream;
using std::string;
#include <iomanip>
using std::setw;
#include <vector>
using std::vector;

int main(int argc, char* argv[]) {
	vector<Token> tokens;
	if (argc != 2) {
		cout << "Usage: " << argv[0] << " [Filename]" << endl;
		return 0;
	}
	if (!initialize(argv[1])) exit(-1);
	while (Token t = get_token()) tokens.push_back(t);
	cleanup();
	//*
	cout << "| Token Type | Token Value" << endl
	     << "+------------+-- - - -  -  -" << endl;
	for (auto& t : tokens) cout << t << endl;
	cout << "+------------+-- - - -  -  -" << endl
	     << endl
	     << endl
	     << "Parsing" << endl;  //*/
	while (tokens.size()) {
		auto ct = parse(tokens);
		cout << "Code tree: " << endl << ct;
		// if (ct.name == "Error") continue;
		// auto prog = generate(ct);
		cout << "Code size: " << /*prog.size()<<*/ " bytes." << endl
		     << "Code execution:" << endl /*<<prog()<<endl*/;
	}
	cout.flush();
}
