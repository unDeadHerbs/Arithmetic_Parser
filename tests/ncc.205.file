# print statement

print( 1-1,1,1+1,1+1+1, 1+1+1+1, 5,2*3,8-1,11-1-1-1,3*3,11-1 );
