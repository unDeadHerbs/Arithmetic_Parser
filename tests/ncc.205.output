Code tree: 
block
  print
    print_i
      -
        1
        1
    print_i
      1
    print_i
      +
        1
        1
    print_i
      +
        +
          1
          1
        1
    print_i
      +
        +
          +
            1
            1
          1
        1
    print_i
      5
    print_i
      *
        2
        3
    print_i
      -
        8
        1
    print_i
      -
        -
          -
            11
            1
          1
        1
    print_i
      *
        3
        3
    print_i
      -
        11
        1
Code size: 338 bytes.
Code execution:
012345678910

