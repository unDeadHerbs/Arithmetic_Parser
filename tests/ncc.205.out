Code tree: 
statement block
  print
    - (sub)
      1
      1
    1
    + (add)
      1
      1
    + (add)
      + (add)
        1
        1
      1
    + (add)
      + (add)
        + (add)
          1
          1
        1
      1
    5
    * (mult)
      2
      3
    - (sub)
      8
      1
    - (sub)
      - (sub)
        - (sub)
          11
          1
        1
      1
    * (mult)
      3
      3
    - (sub)
      11
      1
Code size: 331 bytes.
Code execution:
012345678910

